### ISSUE TITLE
identifying "official" CMIPx replication node downloads

### ISSUE THREAD

**Issue by durack1** [ESGF/esgf-download#32]
Hi all, we've been reviewing the CMIP6 download statistics at http://esgf-ui.cmcc.it/esgf-dashboard-ui/cmip6.html, and am curious about how we could use a custom `user-agent` string to identify "registered" replication nodes (e.g. LLNL/aims3.llnl.gov, DKRZ/esgf1.dkrz.de, CEDA/esgf.ceda.ac.uk, ...) vs "non-registered" general downloads. In the current CMCC stats, these 3 replication nodes account for ~30, 18 and 10% of the CMIP6 downloads through early September 2023.

ping @sandrofiore @alenu @ant4res @sashakames @climate-dude

**Comment by durack1** [comment]
@AtefBN great, thanks for engaging on this. We had thought that have a user-agent that could be configured by a user - the ones we're most interested in are the primary replication users, so for CMIP6/currently we have LLNL, DKRZ, CEDA and NCI - will have to watch how the network evolves going forward. If we could figure out a way to register these instances of esgpull (during the software config), then it *should* be possible for the download statistics to identify "registered" replication services, vs other use cases

**Comment by AtefBN** [comment]
I agree and I don't see why we cannot include this in the software. I am not sure how widely esgpull replaced synda in the replication nodes though, per the CDNOT discussion yesterday, couple nodes are still running on synda which could/should change in the upcoming months.

**Comment by durack1** [comment]
@AtefBN and exactly. At LLNL we are still using (an old version of) Synda, and hadn't planned to update to esgpull until we move to a new project - likely CMIP6Plus preceding CMIP7...

**Comment by alenu** [comment]
Hi all, on the statistics side, we made a test on the "user-agent" field of the HTTP entry we get for each download.
We store the user-agent information, so we can easily filter on a custom value to distinguish between generic downloads and replicas.
Adding @anfab in the thread who is also involved in this activity.

**Comment by AtefBN** [comment]
Hi all, just wanted to revive this conversation to see if there's any specific custom user-agent that should be assigned to the esgpull transfers?

**Comment by durack1** [comment]
@AtefBN if I recall my discussion, we wanted the ESGF replication nodes to identify themselves as "replicators" so that we don't count these as user downloads in the logs.. So what identifier do we want to create that says so. Considering the past nodes who replicated, this will likely include DoE-ORNL, IPSL, CEDA, DKRZ, DIAS. I have not thought much about the CORDEX or other projects, was only really thinking about CMIPx as a first pass.

In the case that the above list works, we could then request that replication nodes configure their esgpull config files to define their specific user-agent, so they can be identified, and then ping the CMCC folks (@alenu, @sandrofiore, @anfab, @sashakames, @mauzey1 ping)

**Comment by AtefBN** [comment]
Thanks @durack1, as things stand esgpull already uses a particular user-agent that should be easy to pick up in traffic "httpx-python3". A brief conversation with @svenrdz yesterday we concluded we can configure a more appropriate custom user-agent if CMCC folks deem it necessary.

**Comment by svenrdz** [comment]
The exact headers used in esgpull are the defaults from [httpx](https://www.python-httpx.org/):
```
│          headers = Headers({'accept': '*/*', 'accept-encoding': 'gzip, deflate', 'connection': 'keep-alive',    │
│                    'user-agent': 'python-httpx/0.28.1'})                                                        │
```

**Comment by durack1** [comment]
@AtefBN @svenrdz ok great, glad to see the specifics.

I would be keen to define a `"user-agent":"python-httpx/0.28.1; DOE-ORNL"` or equivalent for e.g., for other primary replication nodes, which would then allow the CMCC stats to quarantine these as ESGF replication tasks, and not user downloads.

Definitely keen to hear from @alenu, @sandrofiore, @anfab, @sashakames, @mauzey1 to make sure that such an approach could work simply

**Comment by AtefBN** [comment]
I have pinged the replication working team channel on Slack, ideally would be best to discuss this on CDNOT meeting to make a decision whether or not this feature should/would be used through esgpull but we're one month away from the next CDNOT meeting.

@svenrdz seems convinced this is an easy feature to implement as soon as we agree on the user-agent format and the esgpull configuration to identify the replication node. If folks at CMCC are happy with the format to be parsed/consumed we can go ahead with the feature.
